components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/latest/GitpodCopyPortUrlAction.kt

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package io.gitpod.jetbrains.remote.latest
+
+import com.intellij.ide.BrowserUtil
+import com.intellij.notification.NotificationAction
+import com.intellij.notification.NotificationType
+import com.intellij.openapi.actionSystem.ActionPlaces
+import com.intellij.openapi.actionSystem.ActionUpdateThread
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.components.service
+import com.intellij.openapi.ide.CopyPasteManager
+import com.intellij.util.application
+import com.jetbrains.rd.platform.codeWithMe.portForwarding.PortForwardingDataKeys
+import io.gitpod.jetbrains.remote.GitpodManager
+import io.gitpod.supervisor.api.Status.PortsStatusRequest
+import io.gitpod.supervisor.api.StatusServiceGrpc
+import kotlinx.coroutines.launch
+import java.awt.datatransfer.StringSelection
+
+@Suppress("ComponentNotRegistered", "UnstableApiUsage")
+class GitpodCopyPortUrlAction : AnAction() {
+    override fun actionPerformed(e: AnActionEvent) {
+        e.dataContext.getData(PortForwardingDataKeys.PORT)?.let { forwardedPort ->
+            application.coroutineScope.launch {
+                getUrlFromPort(forwardedPort.hostPortNumber)?.let {
+                    CopyPasteManager.getInstance().setContents(StringSelection(it))
+                    service<GitpodManager>()
+                            .notificationGroup
+                            .createNotification("Port URL copied to clipboard!", NotificationType.INFORMATION)
+                            .addAction(NotificationAction.createSimple("Open in browser") {
+                                BrowserUtil.browse(it, e.project)
+                            }).notify(e.project)
+                }
+            }
+        }
+    }
+
+    override fun update(e: AnActionEvent) {
+        // We want it to be enabled only when action comes from the Port Forwarding Suggestion Popup.
+        // In other places, like the Search Actions Menu, we want it disabled (greyed-out).
+        e.presentation.isEnabled = (e.place == ActionPlaces.POPUP)
+    }
+
+    override fun getActionUpdateThread(): ActionUpdateThread {
+        return ActionUpdateThread.BGT
+    }
+
+    private fun getUrlFromPort(port: Number): String? {
```

Comment:
it seems to return expose URL?

What is a semantic of this action? Forwarded if forwarded, otherwise exposed?

if it is only exposed then it should be called Copy Exposed URL, or `Copy Web Link`

Reply:
Shall we have a single "Copy URL" and if the port is forwarded we copy "localhost:xyz" while if the port is not forwarded, we copy the exposed URL? cc @felladrin @akosyakov

Reply:
I think we should have both:
- `Copy URL` -> prefer forwarded if not available use exposed, similar to port notifications
- `Copy Web URL` -> always exposed URL

Reply:
The idea was to display the "exposed url" from Gitpod, as we're not displaying exposed URLs via the New Ports View. Also, the user can already click the Forwarded Port Number (in the screenshot, port 8000) and it will open "localhost:8000".

>I think we should have both:
>
>    Copy URL -> prefer forwarded if not available use exposed, similar to port notifications
>    Copy Web URL -> always exposed URL

Ok, let me update it.

Reply:
Updated!